src/AzureAuth/Commands/Ado/CommandPat.cs

```diff
@@ -63,16 +90,101 @@ private enum OutputMode
         /// </summary>
         /// <param name="logger">The <see cref="ILogger{T}"/> instance that is used for logging.</param>
+        /// <param name="publicClientAuth">An <see cref="IPublicClientAuth"/>.</param>
         /// <param name="eventData">Lasso injected command event data.</param>
         /// <returns>An integer status code. 0 for success and non-zero for failure.</returns>
-        public int OnExecute(ILogger<CommandPat> logger, CommandExecuteEventData eventData)
+        public int OnExecute(ILogger<CommandPat> logger, IPublicClientAuth publicClientAuth, CommandExecuteEventData eventData)
         {
-            logger.LogInformation("coming soon");
+            var accessToken = this.AccessToken(publicClientAuth, eventData);
+            if (accessToken == null)
+            {
+                logger.LogError("Failed to acquire an Azure DevOps access token. Re-run with '--verbosity debug' for more info.");
+                return 1;
+            }
+
+            var patOptions = new PatOptions
+            {
+                Organization = this.Organization,
+                DisplayName = this.DisplayName,
+                Scopes = this.Scopes,
+            };
+
+            var cache = this.Cache();
+            var client = this.Client(accessToken.Token);
+            var manager = new PatManager(cache, client);
+
+            using (new CrossPlatLock(LockfilePath))
+            {
+                var pat = manager.GetPatAsync(patOptions).Result;
+
+                // Do not use logger to avoid printing PATs into log files.
+                Console.WriteLine(FormatPat(pat, this.Output));
+            }
+
             return 0;
         }
+
+        private static string FormatPat(PatToken pat, OutputMode output) => output switch
+        {
+            OutputMode.None => string.Empty,
+            OutputMode.Status => $"\"{pat.DisplayName}\" valid until {pat.ValidTo:O}",
+            OutputMode.Token => pat.Token,
+            OutputMode.Base64 => pat.Token.Base64(),
+            OutputMode.Header => pat.Token.AsHeader(AzureAuth.Ado.Authorization.Basic),
+            OutputMode.HeaderValue => pat.Token.AsHeaderValue(AzureAuth.Ado.Authorization.Basic),
+            OutputMode.Json => pat.AsJson(),
+            _ => throw new ArgumentOutOfRangeException(nameof(output)),
+        };
+
+        private TokenResult AccessToken(IPublicClientAuth publicClientAuth, CommandExecuteEventData eventData)
+        {
+            return publicClientAuth.Token(
+                AzureAuth.Ado.Constants.AdoParams,
+                AccessTokenAuthModes,
+                this.Domain,
+                AccessTokenPrompt,
+                AccessTokenTimeout,
+                eventData);
+        }
+
+        private IPatClient Client(string accessToken)
+        {
+            var baseUrl = new Uri($"{AzureAuth.Ado.Constants.BaseUrl}/{this.Organization}");
+            var credentials = new VssOAuthAccessTokenCredential(accessToken);

```

Comment:
That's neat, I didn't know about `VssOAuthAccessTokenCredential` - we might want to recommend that usage in AzureAuth-Lib docs.

Reply:
I didn't know about it either until I went spelunking in the Visual Studio code. I don't think it's especially well documented. Happy to mention it wherever it seems appropriate.